lsm/main.cpp

```diff
@@ -6,30 +6,169 @@
 #include <vector>
 #include <fstream>
 
+class BloomFilter {
+std::bitset<64> bset;
+public:
+  void insert_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    bset[last_d] = 1;
+  }
+  char contains_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    return bset[last_d];
+  }
+
+};
+
+class File {
+public:
+  std::string filename;
+  int min, max;
+  BloomFilter bloomFilter;
+
+  File(std::string filename) {
+    this->filename = filename;
+    std::vector<int32_t> keys = get_keys();
+    this->min = *min_element(keys.begin(), keys.end());
+    this->max = *max_element(keys.begin(), keys.end());
+    for (auto k : keys) {
+      bloomFilter.insert_key(k);
+    }
+  }
+
+  std::vector<int32_t> get_keys() {
+    std::vector<int32_t> keys;
+    if (auto istrm = std::ifstream(filename, std::ios::binary)) {
+      int k;
+      while (istrm.read(reinterpret_cast<char *>(&k), sizeof k)) {
+        int v;
+        char d;
+        istrm.read(reinterpret_cast<char *>(&v), sizeof v);
+        istrm.read(reinterpret_cast<char *>(&d), sizeof d);
+        keys.push_back(k);
+      }
+    } else {
+      std::cerr << "could not open " << filename << "\n";
+      exit(3);
+    }
+    return keys;
+  }
+};
+
+class maybe_value {
+public:
+  int32_t v;
+  char is_deleted;
+  explicit maybe_value() : v{0}, is_deleted{0} {}
+  explicit maybe_value(int32_t _v, char _deleted) : v{_v}, is_deleted{_deleted} {}
+};
+
 class LSM {
-  std::map<int32_t, int32_t> m;
+  std::map<int32_t, maybe_value> m;
+  std::vector<File> files;
 public:
-  void put(int32_t k, int32_t v) {
-    m.insert(std::make_pair(k,v));
+  void put(int32_t k, int32_t v, char deleted = 0) {
+    if (m.size() >= 4) {
+      std::string filename;
+      if (files.empty()) {
+        filename = "0";
+      } else {
+        filename = std::to_string(std::stoi(files.back().filename) + 1);
+      }
+      write_to_file(filename);
+      files.push_back(File(filename));
+      m.clear();
+      m.insert_or_assign(k,maybe_value(v, deleted));
+    } else {
+      m.insert_or_assign(k,maybe_value(v, deleted));
+    }
   }
   std::optional<int32_t> get(int32_t k) {
     auto it = m.find(k);
-    if (it != m.end()) {
-        return it->second;
+
+    if (it != m.end() && !it->second.is_deleted) {
+      return it->second.v;
+    } else {
+      for (auto file : files) {
+        if (file.bloomFilter.contains_key(k) && k >= file.min && k <= file.max) {
+          std::map<int32_t, maybe_value> file_map = read_from_file(file.filename);
+          auto it_m = file_map.find(k);
+          if (it_m != m.end() && !it_m->second.is_deleted) {
+            return it_m->second.v;
+          }
```

Comment:
You iterate from oldest file to newest file, so I think this will print 3. It should print an empty line.
```
p 1 3
p 10 1
p 11 1
p 12 1
d 1
p 13 1
p 14 1
p 15 1
p 16 1
g 1
```